dxp/demystifying_aui_script_and_liferay-frontend_component_tags.md

```diff
@@ -0,0 +1,114 @@
+
+#### Import an AUI/YUI module using `use` property
+
+When using the `use` property, the AUI / YUI module listed in the` use` list will be imported to the script data. If you want to know a little more about ScriptData, see this [link on how resources are injected into DXP](https://github.com/liferay/liferay-frontend-guidelines/blob/ea185686db8d562ac11cc2b95d2e9e7d7a0c6547/dxp/resource_injection.md#scriptdata).
+
+You can import several YUI / AUI modules within the same `use` property, as in this example:
+
+```jsp
+<aui:script use="liferay-upload">
+...
+        var liferayUpload = new Liferay.Upload({
+...
+</aui:script>
+```
+
+Note in the AlloyUI components, coming from the alloy-ui repository, classes are made available from the global variable `A`, as in this [code snippet the Menu component is defined](https://github.com/liferay/alloy-ui/blob/master/src/aui-menu/js/aui-menu.js#L39]).
+
+However, components made available from the `frontend-js-aui-web` module will be under the global Liferay variable. Like this [example here, `Liferay.Upload`](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-js/frontend-js-aui-web/src/main/resources/META-INF/resources/liferay/upload.js#L1541), used in the code example above.
+
+#### Import an ES6 module using `require` property
+
+Allows users to require a ES6 module as the following example:
+
+Considering that we already have `metal-clipboard` and `metal-dom` dependencies listed [on our module package.json](https://help.liferay.com/hc/en-us/articles/360018159771-The-Structure-of-OSGi-Bundles-Containing-npm-Packages-), We have:
+
+```jsp
+<aui:script require="metal-clipboard/src/Clipboard">
+    new metalClipboardSrcClipboard.default();
+</aui:script>
+```
+
+Also, we can set an alias with the `as` word after the provided path. It will be looking like:
+
+```jsp
+<aui:script require="metal-dom/all/dom.js as dom">
+
+	function myFunction(element) {
+		dom.addClasses(element, 'hide');
+	}
+
+	Liferay.once('something', myFunction);
+
+</aui:script>
+```
+
+#### Settling the way which the script will be placed on the page using `position`:
+
+This property will change the way in which / how the script will be injected into the page. By default, those scripts that are wrapped by `aui:script` will be added to the minified files as described here (Pure JS on the JSP [PUT THE LINK HERE OH]).
+
+When placing the option of `inline` for this property, the scripts will be placed in a script tag obeying the order of the DOM where the JSP is being rendered. See [here](https://github.com/liferay/liferay-portal/blob/master/util-taglib/src/com/liferay/taglib/aui/ScriptTag.java#L143..L165).
+
+#### `sandbox`:
+
+The purpose for the sandbox attribute is to wrap the contents of the script tag in a self-invoking function, and to enable short-handing the `$` and `_` variables rather than calling `AUI.$` and `AUI._`.
```

Comment:
Two questions here:

1. Taking a look on the implementation, We aren't checking if JQuery is available on the page or not. Should I open a ticket for it?

2. What are the benefits in wrapping a script on an anonymous function?

Reply:
2. Wrapping it you're sandboxing it, which means that you are limiting the scope. i.e:
```
<aui:script>
   var myVar = 'foo'; // This variable will be available globally
</aui:script>

<aui:script sandbox="<%= true %>>
   var myVar = 'foo'; // This variable will NOT be available globally
</aui:script>

Reply:
"self-invoking function" is inaccurate. This would be a "self-invoking function" (ie. a recursive one):

```js
function fibonacci(n) {
    if (n <= 1) {
        return 1;
    }

    return fibonacci(n - 1) + fibonacci(n - 2);
}
```

The standard term that I think you're looking for is "IIFE". You don't even need to define it, just make it a link to the MDN [IIFE](https://developer.mozilla.org/en-US/docs/Glossary/IIFE) definition.

> Taking a look on the implementation, We aren't checking if JQuery is available on the page or not. Should I open a ticket for it?

I'd say absolutely not. This is a known issue, but we don't care: jQuery is off by default, and no code that we write from now on will ever rely on it.

So, I think you can drop the references to both `$` and `_`. We do not want to encourage or even admit that jQuery and Underscore exist.

Reply:
> Taking a look on the implementation, We aren't checking if JQuery is available on the page or not. Should I open a ticket for it?

No need. It simply means `A.$ = undefined`. Nothing bad happens for not checking since we do a simple assignment.

Reply:
> I think you can drop the references to both $ and _. We do not want to encourage or even admit that jQuery and Underscore exist.

Makes sense omit this. I'll update

Reply:
> I'd say absolutely not. This is a known issue, but we don't care: jQuery is off by default, and no code that we write from now on will ever rely on it.
> No need. It simply means A.$ = undefined. Nothing bad happens for not checking since we do a simple assignment.

Ok :)